bigquery/storage/managedwriter/client.go

```diff
@@ -0,0 +1,222 @@
+	mw := &ManagedWriteClient{
+		streamSettings: defaultStreamSettings(),
+		client:         client,
+	}
+
+	// apply writer options
+	for _, opt := range opts {
+		opt(mw)
+	}
+
+	if mw.streamSettings.streamID == "" && mw.streamSettings.streamType == "" {
+		return nil, fmt.Errorf("TODO insufficient config?")
+	}
+	if mw.streamSettings.streamID == "" {
+		// not instantiated with a stream, construct one.
+		streamName := fmt.Sprintf("projects/%s/datasets/%s/tables/%s/_default", table.ProjectID, table.DatasetID, table.TableID)
+		if mw.streamSettings.streamType != DefaultStream {
+			// For everything but a default stream, we create a new stream on behalf of the user.
+			req := &storagepb.CreateWriteStreamRequest{
+				Parent: fmt.Sprintf("projects/%s/datasets/%s/tables/%s", table.ProjectID, table.DatasetID, table.TableID),
+				WriteStream: &storagepb.WriteStream{
+					Type: streamTypeToEnum(mw.streamSettings.streamType),
+				}}
+			resp, err := mw.client.CreateWriteStream(ctx, req)
+			if err != nil {
+				return nil, fmt.Errorf("couldn't create write stream: %v", err)
+			}
+			streamName = resp.GetName()
+		}
+		mw.streamSettings.streamID = streamName
+		// TODO(followup CLs): instantiate an appendstream client, flow controller, etc.
+	}
+
+	return mw, nil
+}
+
+func streamTypeToEnum(t StreamType) storagepb.WriteStream_Type {
+	switch t {
+	case CommittedStream:
+		return storagepb.WriteStream_COMMITTED
+	case PendingStream:
+		return storagepb.WriteStream_PENDING
+	case BufferedStream:
+		return storagepb.WriteStream_BUFFERED
+	default:
+		return storagepb.WriteStream_TYPE_UNSPECIFIED
+	}
+}
+
+// StreamName returns the corresponding write stream ID being managed by this writer.
+func (mw *ManagedWriteClient) StreamName() string {
+	return mw.streamSettings.streamID
+}
+
+// StreamType returns the configured type for this stream.
+func (mw *ManagedWriteClient) StreamType() StreamType {
+	return mw.streamSettings.streamType
+}
+
+// FlushRows advances the offset at which rows in a BufferedStream are visible.  Calling
+// this method for other stream types yields an error.
+func (mw *ManagedWriteClient) FlushRows(ctx context.Context, offset int64) (int64, error) {
+	req := &storagepb.FlushRowsRequest{
+		WriteStream: mw.streamSettings.streamID,
+		Offset: &wrapperspb.Int64Value{
+			Value: offset,
+		},
+	}
+	resp, err := mw.client.FlushRows(ctx, req)
+	if err != nil {
+		return 0, err
+	}
+	return resp.GetOffset(), nil
+}
+
+// Finalize is used to mark a stream as complete, and thus ensure no further data can
+// be appended to the stream.  You cannot finalize a DefaultStream, as it always exists.
+//
+// Finalizing does not advance the current offset of a BufferedStream, nor does it commit
+// data in a PendingStream.
+func (mw *ManagedWriteClient) Finalize(ctx context.Context) (int64, error) {
```

Comment:
Once this is called can the client still be used? Should this be a Close for consistency if so?

Reply:
Should client be able to be reused somehow but on a different stream?

Reply:
This shouldn't close a client, because Finalize isn't the end of a stream's interactions.  You still can choose to commit (in the case of pending streams).

There's also some future considerations around whether we do traffic balancing by creating multiple streams under the hood, but that's not currently in scope for this first round.

Reply:
As discussed, latest change breaks it into a Client and a ManagedStream.  ManagedStream retains the settings, Client contains methods that aren't tied to a specific stream (mostly just batch commit streams, and getStream for retreiving stats.